wgpu-core/src/device/mod.rs

```diff
@@ -196,6 +196,30 @@ fn map_buffer<B: hal::Backend>(
         }),
         _ => None,
     };
+    let writes_are_synced = buffer.sync_mapped_writes.is_some() || block.is_coherent();
+
+    // Zero out uninitialized parts of the mapping. (Spec dictates all resources behave as if they were initialized with zero)
+    // Note that going through a `fill_buffer` command may be faster but we don't have the chance of interacting with the queue
+    // between map intent (map_buffer_async or immediately) and this call.
+    let uninitialized_ranges: Vec<Range<BufferAddress>> =
+        buffer.uninitialized_ranges_in_range(Range {
+            start: offset,
+            end: offset + size,
+        });
+    for range in uninitialized_ranges {
+        unsafe {
+            ptr::write_bytes(
+                ptr.as_ptr().offset(range.start as isize),
+                0,
+                (range.end - range.start) as usize,
+            )
+        };
+        // (technically the buffer is only initialized when we're done unmapping but we know it can't be used otherwise meanwhile)
+        if writes_are_synced {
```

Comment:
why are we checking for `writes_are_synced`? what does this variable mean, exactly?

Reply:
When we zero out the user facing memory we naturally can mark the buffer as initialized if it is a write mapping. If it is a read mapping however, this only works if the memory block has the coherent flag. Otherwise, the writes done here are not flushed on unmap and thus may get lost / not be visible to the GPU.
That's what this this boolean captures (I'd check on the fly but borrows don't allow for that).

.. so that's at least how I understood the (un)mapping code around here. Is that accurate?

Reply:
Yes, for the write mapping they should definitely be considered initialized. But *additionally* also for all read mappings on memory that has `HOST_COHERENT`-flag set, because then any changes will be visible to the GPU by definition.
The fact that nothing can use the buffer during the mapping is already used here by doing the mark-as-initialized right away and not in unmap (mostly for convenience).

For reference, `HOST_COHERENT` (== `block.is_coherent()`) in gpu-alloc is speced out as:
```
        /// This flag is set for host-coherent memory types.
        /// Host-coherent memory does not requires manual invalidation for
        /// modifications on GPU to become visible on host;
        /// nor flush for modification on host to become visible on GPU.
        /// Access synchronization is still required.
```

So that's why I put it down as if "either the memory is flushed on unmap (`buffer.sync_mapped_writes.is_some()`) or it is coherent (`|| block.is_coherent()`)", then we can assume that zeroing out the memory is permanent. Otherwise this host sided "change" may not be visible to the GPU so we're not allowed to mark it as initialized.

Does that make sense now or does my argumentation have holes?

Reply:
For all the read mapping, it only comes back to us with a pointer *after* we invalidated the ranges. So regardless of coherent/non-coherent, the data in the buffer that comes in as "mapped for reading" is initialized by that point.

Reply:
> Otherwise this host sided "change" may not be visible to the GPU so we're not allowed to mark it as initialized.

To clarify, we don't need to zero out memory that is read from GPU. The logic you are writing ensures that the source data on GPU is initialized properly, so whatever is being read from GPU to CPU must already contain valid to use data.

Reply:
oh thanks didn't fully digest that mappings are permanent (mentally still stuck a bit in d3d9 style mapping it seems).

> so whatever is being read from GPU to CPU must already contain valid to use data

Hm last mental puzzle piece I'm missing for that is if `block.invalidate_range` won't really be able to drag garbage data from gpu. If I create an uninitialized buffer, map it, get hand-zeroed memory, unmap it, map it again, won't the call to `block.invalidate_range` mean that I could get bad GPU originated memory? After all, the zeros from the first mapping never made it to GPU and now I invalidated my host memory..?

Reply:
can continue this on Matrix if more convenient